### ISSUE TITLE
What is the main difference of asymmetric vs symmetric loss in terms of code?

### ISSUE THREAD

**Issue by sarmientoj24** [mlyg/mixed-focal-loss#16]
Would like to test this on a boundary loss that does not have an asymmetric version. I am trying to implement asymmetry for this loss.
https://github.com/LIVIAETS/boundary-loss/blob/b43fbaf16254abd96ed92244dbe48d4350dde147/losses.py#L76-L93

I am assuming that by the default, Focal Loss (for example), or any losses in general are **symmetric**. Is that correct?

**Comment by mlyg** [comment]
Dear sarmientoj24,

Thank you very much for your question.

You are correct that the default losses you will find are symmetric versions, which means that they affect all classes in the same way. For example, if the focal loss exponentiates the loss by 2, then this will be applied to all classes, although the theory behind the focal loss is that the loss corresponding to the more challenging class will be less downweighted (the more challenging class will have a higher loss -> let's pretend it's 0.9, compared to the easier class, let's say its loss is 0.1 -> exponentiating both losses by 2 leads to 0.9x0.9=0.81 for the challenging class, and 0.1x0.1=0.01 for the easy class i.e. the challenging class is less downweighted).

As for asymmetric losses, the idea is to treat the classes in different ways. With the focal loss, it might seem counterintuitive that we are downweighting the challenging class whose loss we want to preserve. With asymmetric losses, we can instead enhance the challenging class, while simultaneously downweighting the easy class.

As for asymmetry in the boundary loss, I would imagine this would relate to issues with over-segmentation or under-segmentation. With an asymmetrical boundary loss, I suppose either over-segmentation or under-segmentation could be penalised more heavily. I imagine there are use cases for an asymmetric boundary loss where there is systematic over-segmentation or under-segmentation, but I would make sure this is what you are looking for.

As a related aside, I have worked on integrating boundary-based losses to the unified focal loss, and had some better results by applying the boundary loss parameterised by a focal parameter (this modulates the strength of the boundary loss in reference to the deviation from ground truth). You can read the paper here: https://arxiv.org/abs/2111.00534.

Best wishes,

Michael

**Comment by sarmientoj24** [comment]
Hi Michael,

Thank you for the response. I appreciate it.

Thank you for clarifying the main difference between symmetric and asymmetric loss!

> I suppose either over-segmentation or under-segmentation could be penalised more heavily.

This is from the repository, _The boundary loss, at its core, is a pixel-wise multiplication between the network predictions (the softmaxes), and a pre-computed distance map._

It uses some form of distance map. This is how they calculate it.

```
class SurfaceLoss():
    def __init__(self, **kwargs):
        # Self.idc is used to filter out some classes of the target mask. Use fancy indexing
        self.idc: List[int] = kwargs["idc"]
        print(f"Initialized {self.__class__.__name__} with {kwargs}")

    def __call__(self, probs: Tensor, dist_maps: Tensor) -> Tensor:
        assert simplex(probs)
        assert not one_hot(dist_maps)

        pc = probs[:, self.idc, ...].type(torch.float32)
        dc = dist_maps[:, self.idc, ...].type(torch.float32)

        multipled = einsum("bkwh,bkwh->bkwh", pc, dc)

        loss = multipled.mean()

        return loss
```

Do you have any suggestions as to how to implement this asymmetry for this?

 > As a related aside, I have worked on integrating boundary-based losses to the unified focal loss, and had some better results by applying the boundary loss parameterised by a focal parameter (this modulates the strength of the boundary loss in reference to the deviation from ground truth).

Thank you for this. It does look something very similar. I am keen to know what is the main difference between the two since it seems like you are also using some form of "distance" maps

**Comment by mlyg** [comment]
Just to clarify in case the distance map was confusing, the distance map computes the Euclidean distance of each point to the boundary between the foreground and background.

I am not sure if they updated the code since I used it, but I copy below some code from their repository which I used for the surface loss:

```
def calc_dist_map(seg):
    
    # compute distance maps for foreground and background separately
    back_dist = distance(seg[:,:,0])
    fore_dist = distance(seg[:,:,1])

    if (np.max(back_dist) != 0) & (np.max(fore_dist) != 0):
        # normalise distance map for background
        back_dist = back_dist / np.max(back_dist) 
        back_dist = (1-back_dist) * seg[:,:,0] 
        back_dist = np.expand_dims(back_dist, axis=-1)

        # normalise distance map for foreground
        fore_dist = fore_dist / np.max(fore_dist) 
        fore_dist = (1 - fore_dist) * seg[:,:,1]
        fore_dist = np.expand_dims(fore_dist, axis=-1)

        res = np.concatenate([back_dist, fore_dist],axis=-1)

        return res
    else:
        return seg

def calc_dist_map_batch(y_true):
    y_true_numpy = y_true.numpy()
    return np.array([calc_dist_map(y)
                     for y in y_true_numpy]).astype(np.float32)

def surface_loss(y_true, y_pred):
    y_true_dist_map = tf.py_function(func=calc_dist_map_batch,
                                     inp=[y_true],
                                     Tout=tf.float32)
    multiplied = y_pred * y_true_dist_map
    return K.mean(multiplied)
```

I modified the above code to show how we can treat the foreground and background separately, and ss for the SurfaceLoss code you have copied, I imagine something similar would be needed, where the foreground and background probability + distance maps are handled separately (you can see in the SurfaceLoss code that they are treated the same).

I hope this has been helpful, but I apologise for not being able to provide you with the exact code needed to modify the SurfaceLoss for asymmetry.

Best wishes,
Michael
